cardano-cli/src/Cardano/CLI/Json/Friendly.hs

```diff
@@ -288,22 +300,79 @@ friendlyVotingProcedures
   :: ConwayEraOnwards era -> L.VotingProcedures (ShelleyLedgerEra era) -> Aeson.Value
 friendlyVotingProcedures cOnwards x = conwayEraOnwardsConstraints cOnwards $ toJSON x
 
-redeemerIfShelleyBased :: MonadWarning m => CardanoEra era -> TxBody era -> m [Aeson.Pair]
-redeemerIfShelleyBased era tb = monoidForEraInEonA @ShelleyBasedEra era $
-  \shEra -> do
-    redeemerInfo <- friendlyRedeemer shEra tb
-    return ["redeemers" .= redeemerInfo]
-
-friendlyRedeemer :: MonadWarning m => ShelleyBasedEra era -> TxBody era -> m Aeson.Value
-friendlyRedeemer _ (ShelleyTxBody _ _ _ TxBodyNoScriptData _ _) = return Aeson.Null
-friendlyRedeemer _ (ShelleyTxBody _ _ _ (TxBodyScriptData _ _ r) _ _) = encodingToJSON $ L.toCBOR r
+getRedeemerDetails
+  :: forall era m. MonadWarning m => AlonzoEraOnwards era -> TxBody era -> m [Aeson.Pair]
+getRedeemerDetails aeo tb = do
+  let ShelleyTx _ ledgerTx = makeSignedTransaction [] tb
+  redeemerInfo <- friendlyRedeemers ledgerTx
+  return ["redeemers" .= redeemerInfo]
  where
-  encodingToJSON :: MonadWarning m => Encoding -> m Aeson.Value
-  encodingToJSON e =
-    eitherToWarning Aeson.Null $
-      first ("Error decoding redeemer: " ++) $
-        fromFlatTerm (decodeValue True) $
-          toFlatTerm e
+  friendlyRedeemers
+    :: Ledger.Tx (ShelleyLedgerEra era)
+    -> m (Aeson.Value)
+  friendlyRedeemers tx =
+    alonzoEraOnwardsConstraints aeo $ do
+      let redeemerAndExUnitPairList = Map.toList $ Ledger.unRedeemers $ tx ^. Ledger.witsTxL . Ledger.rdmrsTxWitsL
+      redeemerList <- mapM (uncurry $ friendlyRedeemerInfo tx) redeemerAndExUnitPairList
+      return $ Aeson.Array $ Vector.fromList redeemerList
+
+  friendlyRedeemerInfo
+    :: Ledger.Tx (ShelleyLedgerEra era)
+    -> Ledger.PlutusPurpose Ledger.AsIx (ShelleyLedgerEra era)
+    -> (Ledger.Data (ShelleyLedgerEra era), ExUnits)
+    -> m Aeson.Value
+  friendlyRedeemerInfo tx redeemerPurpose (redeemerData, exUnits) =
+    alonzoEraOnwardsConstraints aeo $ do
+      let inputNotFoundError = "Could not find corresponding input to " <> show redeemerPurpose
```

Comment:
Hm. What about putting the warning in the actual JSON output instead of using `MonadWarning`? If we can't find the corresponding input it means the tx is invalid.

Reply:
Addressed in: https://github.com/IntersectMBO/cardano-cli/commit/b7b65f1e09c34ca8a2f5d60d143c44cdbe94b906